programs/survival/spellbook.sc

```diff
@@ -0,0 +1,222 @@
+      if(v != _render_version(book_save),
+        put(block_data, 'Book.tag', nbt(_render_book_nbt(book_save)));
+        without_updates(
+          set(pos, block, {}, block_data);
+        );
+        print(p, str('Automagically Updated %s Spell Book from %s to %s.', book_nbt:'title', v, book_save:'render':'v'));
+      );
+    );
+  );
+);
+
+
+// Standardize reading and writing functions
+_read_book(name) -> (
+  file = read_file(name, 'json');
+  if( file, 
+    return(file);
+  ,
+    file = global_base_book_data;
+    file:'title' = name;
+    return(file);
+  );
+);
+
+_write_book(book) -> (
+  book:'vscript' = global_spellbook_version;
+  book:'vbook' = book:'vbook' + 1;
+  write_file(book:'title', 'json', book);
+);
+
+_write_render(book) -> (
+  write_file(book:'title', 'json', book);
+);
+
+
+
+// Command Methods
+help() -> (
+  print(player(), '
+A utility used to create command books (spell books).
+
+Add or override a spell in a book.
+  /spellbook <book> set <title> <"command">
+  /spellbook farms set "spawn slime farm bot" "/execute in overworld run player SlimeBot spawn at -50.50 82 24"
+  /spellbook farms set "kill slime farm bot" "/player SlimeBot kill"
+  /spellbook farms give
+
+Give the player a spellbook.
+  /spellbook <book> give
+
+Remove a spell.
+  /spellbook <book> remove <spell>
+
+List all spells in a book.
+  /spellbook <book> read
+');
+);
+
+display_book(book_name) -> (
+  book = _read_book(book_name);
+  p = player();
+  print(p, str('%s spells:', book_name));
+  for(pairs(book:'spells'),
+    print(p,str('[ %s ]( %s )', _:0, _:1));
+  );
+);
+
+give_book(name) -> (
+  book = _read_book(name);
+  p = player();
+  print(p, 
+    run(str('/give %s written_book%s', query(p, 'command_name'), _render_book_nbt(book))):1
+  );
+);
+
+delete_command(book_name, spell) -> (
+  book = _read_book(book_name);
+  if( delete(book:'spells':spell),
+    _write_book(book);
+    print(player(), str('Removed the [ %s ] spell.', spell))
+  ,
+    print(player(), str('Unknown spell [ %s ].', spell))
+  );
+);
+
+set_command(book_name, title, command) -> (
+  p = player();
+  book = _read_book(book_name);
+  book:'spells':title = command;
+  print(p, str('%s spell set: [ %s ]( %s ).', book_name, title, command));
+  _write_book(book);
+);
+
+
+
+
+// Book Rendering Functions 
+_render_single_spell(title, command, color) -> (
+  return(str(global_spell_template,title,color,command));
+);
+
+_render_pages(book) -> (
+  spells = pairs(book:'spells');
+  pages = [];
+  a = 0;
+  l = length(spells);
+  while(a<length(spells),50,
+    page = [str(global_title_template, book:'title')];
+    loop( min(global_spells_per_page,l) ,
+      put(page, null, _render_single_spell(spells:a:0, spells:a:1, _shuffle_color()));
+      a += 1;
+    );
+    l = l - global_spells_per_page;
+    put(pages, null, page);
+    if( l < 1, break() );  
+    
+
+  );
+  return(str('[%s]',
+    str('\'[%s]\'', join(']\',\'[', pages)) 
```

Comment:
This line is a little gross in my opinion. each book page is stored as a string but can actually have more nbt structure within it. I am treating the whole nbt tag as a string until its used. 

the other way would be to convert it to nbt and then convert each page back to a string when your done. that would involve some rewriting bits.